{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "053309c2-cb6d-4623-83b7-7b0d947998e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-07-22 12:06:11.515162: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX512F, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[easydl] tensorflow not available!\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "import numpy as np\n",
    "import pytorch_lightning as pl\n",
    "import torchvision.transforms as tf\n",
    "from tqdm import tqdm\n",
    "from predict import *\n",
    "from HIST2ST import *\n",
    "from dataset import ViT_HER2ST, ViT_SKIN\n",
    "from scipy.stats import pearsonr,spearmanr\n",
    "from torch.utils.data import DataLoader\n",
    "from pytorch_lightning.loggers import TensorBoardLogger\n",
    "from copy import deepcopy as dcp\n",
    "from collections import defaultdict as dfd\n",
    "from sklearn.metrics import adjusted_rand_score as ari_score\n",
    "from sklearn.metrics.cluster import normalized_mutual_info_score as nmi_score\n",
    "import os\n",
    "\n",
    "os.environ['CUDA_LAUNCH_BLOCKING'] = '1'  # For debugging CUDA errors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b132bfb4-bd29-44fa-bbca-3790b0848fd4",
   "metadata": {},
   "source": [
    "# Data Loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6281d8f2-e5f7-4c26-88bf-5256197b1418",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Generate a list of names for different groups (A-G) with specified ranges\n",
    "name = [*[f'A{i}' for i in range(2,7)], *[f'B{i}' for i in range(1,7)],\n",
    "    *[f'C{i}' for i in range(1,7)], *[f'D{i}' for i in range(1,7)],\n",
    "    *[f'E{i}' for i in range(1,4)], *[f'F{i}' for i in range(1,4)], *[f'G{i}' for i in range(1,4)]]\n",
    "\n",
    "# List of patient identifiers\n",
    "patients = ['P2', 'P5', 'P9', 'P10']\n",
    "\n",
    "# List of replicate identifiers\n",
    "reps = ['rep1', 'rep2', 'rep3']\n",
    "\n",
    "# Generate skin sample names by combining patient and replicate info\n",
    "skinname = []\n",
    "for i in patients:\n",
    "    for j in reps:\n",
    "        skinname.append(i + '_ST_' + j)\n",
    "\n",
    "# Set device for computation\n",
    "device = 'cuda'\n",
    "\n",
    "# Model configuration tag (used for hyperparameters)\n",
    "tag = '5-7-2-8-4-16-32'\n",
    "\n",
    "# Unpack model hyperparameters from tag string\n",
    "k, p, d1, d2, d3, h, c = map(lambda x: int(x), tag.split('-'))\n",
    "\n",
    "# Set dropout rate for model\n",
    "dropout = 0.2\n",
    "\n",
    "# Set random seeds for reproducibility\n",
    "random.seed(12000)\n",
    "np.random.seed(12000)\n",
    "torch.manual_seed(12000)\n",
    "torch.cuda.manual_seed(12000)\n",
    "torch.cuda.manual_seed_all(12000)\n",
    "torch.backends.cudnn.benchmark = False\n",
    "torch.backends.cudnn.deterministic = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5674163",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionaries to convert between Ensembl gene IDs and HGNC symbols\n",
    "conversion = pd.read_csv('../../../../tahsin/HEST/assets/gene_ids/hgnc_complete_set.txt', sep=\"\\t\").loc[:, ['ensembl_gene_id', 'symbol']]\n",
    "convert_to_ens = conversion.set_index('symbol')['ensembl_gene_id'].to_dict()\n",
    "convert_to_sym = conversion.set_index('ensembl_gene_id')['symbol'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1151fa5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def emb_to_sym(label):\n",
    "    \"\"\"\n",
    "    Convert the label to a list of gene symbols.\n",
    "    \"\"\"\n",
    "    if isinstance(label, str):\n",
    "        for gene in label.split(' '):\n",
    "            if gene not in convert_to_sym and gene in convert_to_ens:\n",
    "                # print(f\"Already a symbol.\")\n",
    "                return label\n",
    "            elif gene not in convert_to_sym:\n",
    "                print(f\"Error: {gene} not in conversion dictionary.\")\n",
    "            else:\n",
    "                conversion = [convert_to_sym[gene] for gene in label.split(' ')]\n",
    "                return ' '.join(conversion)\n",
    "    elif isinstance(label, list):\n",
    "        conversion = [convert_to_sym[gene] for gene in label]\n",
    "        return ' '.join(conversion)\n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a0c94a0",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "\n",
    "# Hist2ST Prediction\n",
    "\n",
    "### To run the trained model, please select the trained model and replace the value of the variable fold with the number in the name of the selected trained model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0f19d5d8-6510-493b-9287-ba7215155dc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fold=5\n",
    "data='her2st'\n",
    "# data = 'hest1k'\n",
    "if data == 'her2st':\n",
    "    prune = 'Grid'\n",
    "    genes = 785\n",
    "elif data == 'cscc':\n",
    "    prune = 'NA'\n",
    "    genes = 171\n",
    "elif data == 'hest1k':\n",
    "    prune = 'NA'# IDK??\n",
    "    genes = 785\n",
    "prune='Grid' if data=='her2st' else 'NA'\n",
    "genes=171 if data=='cscc' else 785"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d3554f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "hest_path = \"/work/bose_lab/tahsin/data/HEST\"\n",
    "\n",
    "meta_df = pd.read_csv(os.path.join(hest_path, \"HEST_v1_1_0.csv\"))\n",
    "meta_df = meta_df[meta_df['species'] == 'Homo sapiens']\n",
    "\n",
    "\n",
    "# Filter the dataframe if you need\n",
    "# meta_df = meta_df[(meta_df['organ'] == 'Heart') & (meta_df['species'] == 'Homo sapiens')]\n",
    "# print(meta_df.shape)\n",
    "\n",
    "# Separate dataframe into training and validation sets\n",
    "# meta_df = meta_df.sample(frac=1, random_state=42)  # Shuffle the dataframe\n",
    "# train_frac = 0.8\n",
    "# train_size = int(len(meta_df) * train_frac)\n",
    "# train_df_80 = meta_df.iloc[:train_size]\n",
    "# val_df_20 = meta_df.iloc[train_size:]\n",
    "\n",
    "ids = meta_df['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d65161f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HPS6' 'TNC' 'NR1H2' 'NUP93' 'HNRNPUL2' 'MARS' 'SUGP1' 'DEAF1' 'WDR13'\n",
      " 'FLNA' 'ELN' 'MX2' 'COQ4' 'LARP7' 'TXNDC17' 'MIIP' 'AQP1' 'POSTN' 'PSCA'\n",
      " 'FAM117A' 'AKT1S1' 'DVL1' 'SPIN1' 'LUC7L' 'TTC31' 'CIR1' 'GAR1'\n",
      " 'RALGAPA2' 'TIMP1' 'GNAI2' 'WDR73' 'UXS1' 'CNN2' 'C1QB' 'CCDC106'\n",
      " 'ARHGAP10' 'ATL2' 'UBL7' 'NDUFA12' 'PTRF' 'PPP1R1A' 'KAT6B' 'CORO1A'\n",
      " 'EZH2' 'CSF1' 'CD79A' 'SDHB' 'GATC' 'TMEM160' 'UXT' 'ZNF787' 'ATF6B'\n",
      " 'SNRPD3' 'ZNF217' 'SLC35A2' 'MMP9' 'ECHDC2' 'PEX11G' 'SF3B4' 'SYDE1'\n",
      " 'IGKC' 'RASGRP2' 'RND3' 'PNISR' 'HSPB1' 'UBR2' 'TESK1' 'STAMBP' 'DPYSL3'\n",
      " 'CPSF1' 'ZMYND19' 'COMP' 'TBCA' 'PPHLN1' 'MYL6B' 'AHCY' 'TRAP1' 'R3HDM2'\n",
      " 'TMEM123' 'TOM1L2' 'VPS13D' 'CD3D' 'MID1IP1' 'DPP7' 'LTBP3' 'GANAB'\n",
      " 'HLA-DOA' 'FAM193B' 'BRPF1' 'ITGB6' 'GBP5' 'CADM4' 'ARID1B' 'CEP250'\n",
      " 'PTRH1' 'GPSM1' 'DLG1' 'MLLT4' 'APBB2' 'C19orf24' 'CCS' 'DHX16' 'CRACR2B'\n",
      " 'CRABP2' 'MEN1' 'ARHGEF40' 'NDUFV3' 'HMGB2' 'ZFPL1' 'POLR2E' 'DAPK3'\n",
      " 'FAM213B' 'RNF135' 'FDPS' 'GPS1' 'DDAH1' 'CSF1R' 'CHCHD6' 'ARID3A'\n",
      " 'RABGAP1L' 'PHF11' 'IDH3G' 'SLC43A1' 'PER1' 'CORO7' 'DNAJC1' 'KMT2B'\n",
      " 'PSMD4' 'ADCY3' 'CTU2' 'PODXL' 'HLA-DMA' 'ATP6V0D1' 'BGN' 'GNAS' 'STRN3'\n",
      " 'PAQR4' 'MPND' 'METTL12' 'SLC9A3R2' 'CLDN3' 'TBXAS1' 'EFHD2' 'APOBEC3G'\n",
      " 'LRWD1' 'HLA-DPA1' 'ZNF592' 'ZG16B' 'NIT2' 'INPP4B' 'C4orf48' 'CPVL'\n",
      " 'TCEA2' 'CCDC167' 'MUCL1' 'MGMT' 'AKAP9' 'TXN2' 'YIPF4' 'RASSF3' 'HSPA4'\n",
      " 'SGTA' 'MLST8' 'MFAP4' 'FAIM2' 'TM7SF2' 'EXD3' 'SLFN13' 'ARID5B' 'EPHB3'\n",
      " 'CCL2' 'FAM118A' 'SMG9' 'MS4A6A' 'PIGQ' 'KLF9' 'ITGA5' 'C1S' 'NANS'\n",
      " 'AGPAT2' 'MST1R' 'ARF5' 'S100A14' 'SEMA3B' 'LMF1' 'ZFP36' 'CHFR' 'CFDP1'\n",
      " 'DDB2' 'AVEN' 'TDRD3' 'RAB11FIP1' 'SLC22A18' 'SRSF5' 'DNAJC7' 'NTMT1'\n",
      " 'SLC4A5' 'HLA-B' 'EXOSC2' 'MVP' 'NUP214' 'VIM' 'PTGDS' 'COL6A2' 'CPNE1'\n",
      " 'PDCD5' 'IGLC7' 'EXOSC10' 'TSR2' 'ATG12' 'ALG3' 'HYI' 'CCNE2' 'JAK1'\n",
      " 'ZMYND8' 'UBAP2L' 'C5orf38' 'JCHAIN' 'FNBP4' 'TRIM22' 'DPP9' 'NDUFA1'\n",
      " 'MORC2' 'MIF' 'IL10RA' 'ATP5O' 'RASA2' 'SCD' 'KCTD10' 'NDUFS1' 'UCK1'\n",
      " 'FUOM' 'CD52' 'MED25' 'RSL1D1' 'TMEM259' 'NEK7' 'FAP' 'PPM1G' 'SPSB1'\n",
      " 'KRT17' 'S100A9' 'BABAM1' 'ZNF771' 'RNASET2' 'IGHM' 'STAT5A' 'COX5A'\n",
      " 'KPNB1' 'MKNK1' 'C19orf52' 'ADI1' 'AGO2' 'RGS1' 'IFI27L1' 'MAPKAPK5'\n",
      " 'LYL1' 'MUS81' 'CAMTA2' 'GBP1' 'ASL' 'TIGD5' 'TMEM101' 'HOOK3' 'KMT2E'\n",
      " 'EIF3B' 'ELOF1' 'MZT2A' 'GPR108' 'KIAA1211L' 'MCTS1' 'TYROBP' 'C19orf60'\n",
      " 'ACER3' 'RAB3GAP1' 'C3' 'PLVAP' 'SURF1' 'LIMCH1' 'P4HA2' 'WDR34'\n",
      " 'PRPSAP1' 'NXN' 'FKBP10' 'PTOV1' 'MED13L' 'RBM42' 'BOK' 'C12orf45'\n",
      " 'SPAG9' 'BCAM' 'FKBP8' 'SNAPC3' 'ZNF791' 'AP5B1' 'GPRC5A' 'TAX1BP1'\n",
      " 'MICAL1' 'ZNF337' 'PTPRC' 'RALBP1' 'ZMAT5' 'GSTM3' 'PHF21A' 'CD14'\n",
      " 'PDGFRB' 'PRPF6' 'SRGN' 'MRPL2' 'CES2' 'FSTL3' 'IGHA1' 'SCAF1' 'TRIM47'\n",
      " 'MRPL51' 'XPO6' 'SOS2' 'POLR2J' 'MAP1S' 'PNMT' 'SP6' 'SIGMAR1' 'DNAJB2'\n",
      " 'SASH1' 'EVA1B' 'CDK5RAP1' 'IKZF3' 'BIRC2' 'MGP' 'IFI44L' 'SNTA1' 'PWP1'\n",
      " 'DCAF6' 'EIF1AD' 'SAT1' 'BCL2L12' 'ASNS' 'RMND1' 'SRSF1' 'ABCA7' 'CTDSPL'\n",
      " 'KIF13B' 'NMRAL1' 'POR' 'MRPL54' 'MAP3K14' 'GPX3' 'ARHGEF10L' 'TSPAN33'\n",
      " 'PCIF1' 'SNX8' 'ZFAND3' 'COPS7A' 'HES1' 'PDGFRA' 'ETFB' 'PSTPIP1' 'MYO1G'\n",
      " 'CCT4' 'ABCA3' 'CDH11' 'COL4A1' 'RBL1' 'MIB2' 'PODNL1' 'PLK1' 'FASN'\n",
      " 'AP5Z1' 'RHOB' 'SCX' 'LAMTOR2' 'PPP1R37' 'BAIAP3' 'ABHD14A' 'AEBP1'\n",
      " 'MAGOHB' 'XIAP' 'SLC2A1' 'TEX10' 'MAPK9' 'ISG15' 'TPM2' 'AAMDC'\n",
      " 'KIAA1715' 'PKNOX1' 'HLA-DPB1' 'PYURF' 'TARBP1' 'A2M' 'SOX4' 'CCDC80'\n",
      " 'SMYD2' 'PUS3' 'CORO1B' 'LST1' 'SEMA4B' 'ENPP1' 'IGFBP2' 'PIM2' 'SSSCA1'\n",
      " 'PLSCR1' 'GMFG' 'GEMIN4' 'BSG' 'EBAG9' 'VWF' 'FGFR1' 'ANKRD28' 'RECQL5'\n",
      " 'CGGBP1' 'ARHGEF3' 'TMEM86A' 'EXOSC1' 'ZSWIM8' 'HDAC6' 'ANKLE2' 'TRAPPC4'\n",
      " 'ARNT' 'PILRA' 'KIAA1217' 'HDAC5' 'TFF3' 'MFSD7' 'FAM173A' 'TNNT1'\n",
      " 'DOCK6' 'ANGPTL4' 'THOC6' 'ARFGEF2' 'COMMD1' 'SCNN1D' 'CLDN4' 'ERI3'\n",
      " 'KRT86' 'LAMB2' 'FUS' 'METTL16' 'PIH1D1' 'HCFC1R1' 'KCNAB2' 'PTPRJ'\n",
      " 'LARP1B' 'PAF1' 'MYH10' 'ABCF3' 'GLTSCR2' 'REEP6' 'IGHG3' 'C10orf54'\n",
      " 'CX3CL1' 'IDUA' 'PLK2' 'SAMM50' 'NDUFB9' 'STK17A' 'CTSK' 'TANC2' 'MAP3K6'\n",
      " 'ERAL1' 'MPST' 'NAA10' 'RAMP2' 'PTP4A3' 'LINGO1' 'EMC3' 'PLAC9' 'EXOC5'\n",
      " 'CRELD2' 'PMP22' 'PLTP' 'ZNF524' 'NCKAP1L' 'APOC1' 'PPP1R1B' 'PLIN2'\n",
      " 'ZNF362' 'THOC5' 'GLG1' 'AKR1A1' 'PPIB' 'BRK1' 'VCAM1' 'AGAP2' 'SEC23A'\n",
      " 'SLC9A8' 'PRKAB1' 'IGHG4' 'HCST' 'PEX14' 'SYK' 'TRIM25' 'PWWP2B' 'FSTL1'\n",
      " 'RIOK1' 'FZD2' 'FN1' 'GSK3A' 'BRF1' 'VPS45' 'FAM102A' 'TRAF5' 'JAG2'\n",
      " 'PCCB' 'TTI1' 'KRT10' 'ZNF91' 'TMX1' 'IGLC2' 'ARID4A' 'ASCL2' 'RSRP1'\n",
      " 'GBP3' 'UBE2W' 'FARP1' 'ATAD1' 'DDT' 'ATG16L2' 'RAB3A' 'LINC00176' 'XAB2'\n",
      " 'PLXND1' 'ARHGAP5' 'GRK3' 'CCL19' 'MYO9B' 'BOLA3' 'NSMCE1' 'UQCRC1'\n",
      " 'VARS' 'CCDC57' 'TEX2' 'CCL21' 'TXNRD2' 'TESMIN' 'TMEM14B' 'CD320'\n",
      " 'RAP1B' 'ABHD4' 'PXDN' 'CHST11' 'OSBPL8' 'PAFAH1B3' 'DPH3' 'FAM45A'\n",
      " 'SLPI' 'EIF3G' 'SOD3' 'ASNA1' 'FBF1' 'SFSWAP' 'GBP4' 'HELZ2' 'STT3A'\n",
      " 'ITGB4' 'ASMTL' 'NDUFB2' 'GYPC' 'PLXNA1' 'PLLP' 'TJP3' 'APBA2' 'COL3A1'\n",
      " 'NRP1' 'CXCL10' 'ERMP1' 'AGR2' 'FAAH' 'CCM2' 'INPP5J' 'FAH' 'THBS1'\n",
      " 'NFIC' 'PLP2' 'GLCCI1' 'ZNF644' 'HSPB11' 'SPARC' 'NEBL' 'ATP5H'\n",
      " 'KIAA0895L' 'TMBIM6' 'SNF8' 'NDUFB3' 'DDIT3' 'KATNA1' 'BAD' 'WDR74'\n",
      " 'PROSER1' 'KIAA1468' 'FBLN2' 'INPP5D' 'ECM1' 'CYP27A1' 'VEGFB' 'C1R'\n",
      " 'KRAS' 'FCER1G' 'MUC1' 'MTMR12' 'PNPLA6' 'CUX1' 'FOXP4' 'ZDHHC12'\n",
      " 'C10orf10' 'ZNF33B' 'IFI27' 'TNFRSF21' 'CBLC' 'ARHGAP4' 'WIPF1' 'RAB3IL1'\n",
      " 'C6orf48' 'ASH2L' 'MRPL21' 'KDM4B' 'TCF25' 'C6orf136' 'COL8A2' 'RARS2'\n",
      " 'RNF170' 'WDR46' 'ILKAP' 'TBC1D10A' 'BRF2' 'BST2' 'BRD9' 'OASL' 'CHD1'\n",
      " 'GAS6' 'OPTN' 'DMAP1' 'CORO2A' 'DPM2' 'C2CD2L' 'UBA6' 'TBC1D23' 'LUC7L3'\n",
      " 'JAK3' 'CALML5' 'LRRC8A' 'ATP6AP1' 'PHF13' 'LCP1' 'MAVS' 'MFSD3' 'TRIM14'\n",
      " 'ALDH16A1' 'INTS8' 'LUM' 'ZNF83' 'ARF6' 'PARN' 'ISLR' 'GBP2' 'WDFY3'\n",
      " 'ADAM15' 'MAP4K2' 'MRPL23' 'C2' 'CD4' 'KRT8' 'CUTC' 'HAGHL' 'ACAA1' 'ZFR'\n",
      " 'LANCL1' 'SLC8B1' 'TRAC' 'ANO10' 'SIGLEC1' 'OGFRL1' 'BIN3' 'NDUFC1'\n",
      " 'RRAS' 'GATA3' 'CCNB1IP1' 'IER5L' 'DERL3' 'PNP' 'TACC3' 'CERCAM' 'NINJ1'\n",
      " 'C1orf216' 'TCEA3' 'TMEM184A' 'BICDL1' 'DCXR' 'NELFE' 'FARSB' 'PPIH'\n",
      " 'CERS5' 'GMPPA' 'SH3BP5L' 'ZNF668' 'FADS2' 'NR4A1' 'COL14A1' 'PRKCSH'\n",
      " 'HLA-DRA' 'PLEKHO2' 'NCKAP5L' 'LIMK1' 'ITSN1' 'ARID5A' 'STMN1' 'FAM3A'\n",
      " 'TNFRSF4' 'GALK1' 'MYL9' 'SVIL' 'CLU' 'LCK' 'SMCHD1' 'QSOX1' 'STK4'\n",
      " 'IGLC3' 'JMJD6' 'RUVBL1' 'CFAP36' 'EFEMP2' 'FAM160B1' 'SLC16A2' 'SNX33'\n",
      " 'MAGEF1' 'DENND1A' 'PITX1' 'UFD1L' 'NFATC4' 'ORC2' 'CROCC' 'AES'\n",
      " 'HLA-DMB' 'ANKRD12' 'UBA52' 'SEC14L2' 'PLPPR2' 'CD74' 'CTIF' 'AURKAIP1'\n",
      " 'LGALS1' 'MTG2' 'TAX1BP3' 'SLC25A24' 'DESI1' 'RECQL4' 'CCDC88B' 'MYL12B'\n",
      " 'SRRT' 'INAFM1' 'LTB4R' 'CARS2' 'DDRGK1' 'CHD6' 'HSD17B1' 'YTHDC1'\n",
      " 'C15orf61' 'CAPG' 'PHKB' 'HLA-DRB1' 'BTG1' 'LMNA' 'SRPX2' 'MRPS22'\n",
      " 'CXXC5' 'PGLS' 'LTF' 'ZNRF1' 'IGFBP4' 'CLASRP' 'ARHGAP45' 'MICALL2'\n",
      " 'MAGI1' 'SRR' 'CD68' 'GNL2' 'TP53I3' 'TGFBR2' 'UBXN1' 'EIF3F' 'SLC2A8'\n",
      " 'LLGL1' 'SLC16A3' 'VAV1' 'NDRG1' 'TNFRSF14']\n"
     ]
    }
   ],
   "source": [
    "gene_list = list(np.load('data/her_hvg_cut_1000.npy',allow_pickle=True))\n",
    "print(np.array(gene_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c4cd9252",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_lists = []\n",
    "# list_o_lists = []\n",
    "# for id in ids:\n",
    "#     adata_path = os.path.join(hest_path, \"st\", id + \".h5ad\")\n",
    "#     if os.path.exists(adata_path):\n",
    "#         adata = sc.read_h5ad(adata_path)\n",
    "#         print(f\"Loaded {id} with shape {adata.shape}\")\n",
    "#         gene_list = adata.var_names.tolist()\n",
    "#         gene_list = [emb_to_sym(gene) for gene in gene_list if gene.startswith(\"ENSG\")]\n",
    "#         str = ' '.join(gene_list)\n",
    "#         gene_lists.append(str)\n",
    "#         list_o_lists.append(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "59b60aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import importlib\n",
    "# import predict\n",
    "# importlib.reload(predict)\n",
    "# from predict import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "48a0e1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import importlib\n",
    "# import dataset\n",
    "# importlib.reload(dataset)\n",
    "# from dataset import ViT_HEST1K\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "76cfcfb9-3169-46da-9288-2e2be7942d84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['B1']\n",
      "Loading imgs...\n",
      "Loading metadata...\n",
      "X embedding dimension: 64\n",
      "Y embedding dimension: 64\n"
     ]
    }
   ],
   "source": [
    "testset = pk_load(fold,'test',dataset=data,flatten=False,adj=True,ori=True,prune=prune)\n",
    "test_loader = DataLoader(testset, batch_size=1, num_workers=0, shuffle=False)\n",
    "# label=testset.label[testset.names[0]]\n",
    "label = meta_df[meta_df['id'] == testset.names[0]]['disease_state']\n",
    "genes=785\n",
    "model=Hist2ST(\n",
    "    depth1=d1, depth2=d2,depth3=d3,n_genes=genes, \n",
    "    kernel_size=k, patch_size=p,\n",
    "    heads=h, channel=c, dropout=0.2,\n",
    "    zinb=0.25, nb=False,\n",
    "    bake=5, lamb=0.5, \n",
    ")\n",
    "model.load_state_dict(torch.load(f'./downloads/{fold}-Hist2ST.ckpt'))\n",
    "print(f\"X embedding dimension: {model.x_embed.num_embeddings}\")\n",
    "print(f\"Y embedding dimension: {model.y_embed.num_embeddings}\")\n",
    "sample_data = testset[0] \n",
    "# patches, positions, expression, adj, centers = sample_data\n",
    "# print(f\"Position coordinate range: {positions.min()} to {positions.max()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "86693610",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current params: k=5, p=7, d1=2, d2=8, d3=4, h=16, c=32\n",
      "patch_size=7, channel=32\n"
     ]
    }
   ],
   "source": [
    "# Your current parameters from tag '5-7-2-8-4-16-32'\n",
    "print(f\"Current params: k={k}, p={p}, d1={d1}, d2={d2}, d3={d3}, h={h}, c={c}\")\n",
    "print(f\"patch_size={p}, channel={c}\")\n",
    "\n",
    "# Check what the model actually expects vs. receives\n",
    "sample_data = testset[0]\n",
    "# patches, positions, expression, adj, centers = sample_data\n",
    "# print(f\"Input patch shape: {patches.shape}\")\n",
    "# print(f\"Model expects channel dimension: {c}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6d691123",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checkpoint keys (first 20):\n",
      " 1: patch_embedding.weight\n",
      " 2: patch_embedding.bias\n",
      " 3: x_embed.weight\n",
      " 4: y_embed.weight\n",
      " 5: vit.transformer.layer1.0.dw.0.weight\n",
      " 6: vit.transformer.layer1.0.dw.0.bias\n",
      " 7: vit.transformer.layer1.0.dw.1.weight\n",
      " 8: vit.transformer.layer1.0.dw.1.bias\n",
      " 9: vit.transformer.layer1.0.dw.1.running_mean\n",
      "10: vit.transformer.layer1.0.dw.1.running_var\n",
      "11: vit.transformer.layer1.0.dw.1.num_batches_tracked\n",
      "12: vit.transformer.layer1.0.dw.3.weight\n",
      "13: vit.transformer.layer1.0.dw.3.bias\n",
      "14: vit.transformer.layer1.0.dw.4.weight\n",
      "15: vit.transformer.layer1.0.dw.4.bias\n",
      "16: vit.transformer.layer1.0.dw.4.running_mean\n",
      "17: vit.transformer.layer1.0.dw.4.running_var\n",
      "18: vit.transformer.layer1.0.dw.4.num_batches_tracked\n",
      "19: vit.transformer.layer1.0.pw.0.weight\n",
      "20: vit.transformer.layer1.0.pw.0.bias\n",
      "... total keys: 162\n"
     ]
    }
   ],
   "source": [
    "# Load and inspect the checkpoint\n",
    "checkpoint = torch.load(f'./downloads/{fold}-Hist2ST.ckpt')\n",
    "\n",
    "print(\"Checkpoint keys (first 20):\")\n",
    "for i, key in enumerate(checkpoint.keys()):\n",
    "    print(f\"{i+1:2d}: {key}\")\n",
    "    if i >= 19:\n",
    "        break\n",
    "print(f\"... total keys: {len(checkpoint.keys())}\")\n",
    "\n",
    "# If the checkpoint contains metadata about architecture\n",
    "if 'hyper_parameters' in checkpoint:\n",
    "    print(\"\\nSaved hyperparameters:\")\n",
    "    for k, v in checkpoint['hyper_parameters'].items():\n",
    "        print(f\"  {k}: {v}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5d88f8ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A4']\n",
      "Loading imgs...\n",
      "Loading metadata...\n"
     ]
    }
   ],
   "source": [
    "test_set = pk_load(2, 'test', dataset=data, flatten=False, adj=True, ori=True, prune=prune)\n",
    "test2_loader = DataLoader(test_set, batch_size=1, num_workers=0, shuffle=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7850bde8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample with label: torch.Size([1, 343, 3, 112, 112])\n",
      "Processing sample with label: torch.Size([1, 343, 2])\n",
      "Processing sample with label: torch.Size([1, 343, 785])\n",
      "Processing sample with label: torch.Size([1, 343, 343])\n",
      "Processing sample with label: torch.Size([1, 343, 785])\n",
      "Processing sample with label: torch.Size([1, 343])\n",
      "Processing sample with label: torch.Size([1, 343, 2])\n",
      "=====\n"
     ]
    }
   ],
   "source": [
    "for d in iter(test2_loader):\n",
    "    for k in d:\n",
    "        print(f\"Processing sample with label: {k.shape}\")\n",
    "    print(\"=====\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "194141e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample with label: torch.Size([1, 295, 3, 112, 112])\n",
      "Processing sample with label: torch.Size([1, 295, 2])\n",
      "Processing sample with label: torch.Size([1, 295, 785])\n",
      "Processing sample with label: torch.Size([1, 295, 295])\n",
      "Processing sample with label: torch.Size([1, 295, 785])\n",
      "Processing sample with label: torch.Size([1, 295])\n",
      "Processing sample with label: torch.Size([1, 295, 2])\n",
      "=====\n"
     ]
    }
   ],
   "source": [
    "for d in iter(test_loader):\n",
    "    for k in d:\n",
    "        print(f\"Processing sample with label: {k.shape}\")\n",
    "    print(\"=====\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "712e84ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_ids = meta_df['id'].tolist()\n",
    "np.random.seed(42)\n",
    "np.random.shuffle(all_ids)\n",
    "\n",
    "split_idx = int(len(all_ids) * 0.8)\n",
    "sample_ids = all_ids[split_idx:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "804e3778",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for sample_id in sample_ids:\n",
    "#     hestset = pk_load(fold, 'test', sample_ids=[sample_id], dataset='hest1k', flatten=False, adj=True, ori=True, prune=prune)\n",
    "#     hest_loader = DataLoader(hestset, batch_size=1, num_workers=0, shuffle=False)\n",
    "#     for d in iter(hest_loader):\n",
    "#         for k in d:\n",
    "#             print(f\"Processing HEST sample with label: {k.shape}\")\n",
    "#         print(\"=====\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7d8b6e64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "dataset_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "image_filename",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "organ",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "disease_state",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "oncotree_code",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "species",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "patient",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "st_technology",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "data_publication_date",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "license",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "study_link",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "download_page_link1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "inter_spot_dist",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "spot_diameter",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "spots_under_tissue",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "preservation_method",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "nb_genes",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "treatment_comment",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "pixel_size_um_embedded",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "pixel_size_um_estimated",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "magnification",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fullres_px_width",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "fullres_px_height",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "tissue",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "disease_comment",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "subseries",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "hest_version_added",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "e9145ec2-c07a-4f71-93e5-96e54390a98d",
       "rows": [
        [
         "1",
         "FFPE Human Skin Primary Dermal Melanoma with 5K Human Pan Tissue and Pathways Panel",
         "TENX158",
         "TENX158.tif",
         "Skin",
         "Cancer",
         "SKCM",
         "Homo sapiens",
         null,
         "Xenium",
         "7/31/24",
         "Creative Commons Attribution",
         null,
         "https://www.10xgenomics.com/datasets/xenium-prime-ffpe-human-skin",
         null,
         null,
         "2179",
         "FFPE",
         "10017",
         null,
         "0.2737769309918141",
         "0.2737536812559247",
         "40x",
         "18669",
         "35787",
         "Skin",
         "Primary Dermal Melanoma ",
         null,
         "v1_1_0"
        ],
        [
         "2",
         "FFPE Human Prostate Adenocarcinoma with 5K Human Pan Tissue and Pathways Panel",
         "TENX157",
         "TENX157.tif",
         "Prostate",
         "Cancer",
         "PRAD",
         "Homo sapiens",
         null,
         "Xenium",
         "7/31/24",
         "Creative Commons Attribution",
         null,
         "https://www.10xgenomics.com/datasets/xenium-prime-ffpe-human-prostate",
         null,
         null,
         "4427",
         "FFPE",
         "10006",
         null,
         "0.2737717292963757",
         "0.2737414653044506",
         "40x",
         "25002",
         "49976",
         "Prostate",
         null,
         null,
         "v1_1_0"
        ],
        [
         "3",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX156",
         "TENX156.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Visium HD",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         "2.0",
         "2.0",
         "2179",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.273802324565481",
         "40x",
         "71106",
         "58791",
         "Colon",
         "Stage II-A ",
         "Visium HD, Sample P1 CRC",
         "v1_1_0"
        ],
        [
         "4",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX155",
         "TENX155.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Visium HD",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         "2.0",
         "2.0",
         "2281",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.2738744199268744",
         "40x",
         "75250",
         "48740",
         "Colon",
         null,
         "Visium HD, Sample P2 CRC",
         "v1_1_0"
        ],
        [
         "5",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX154",
         "TENX154.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Visium HD",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         "2.0",
         "2.0",
         "2220",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.273770509329827",
         "40x",
         "72897",
         "64370",
         "Colon",
         null,
         "Visium HD, Sample P5 CRC",
         "v1_1_0"
        ],
        [
         "6",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX153",
         "TENX153.tif",
         "Bowel",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 1",
         "Visium HD",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         "2.0",
         "2.0",
         "2018",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.273906543053634",
         "40x",
         "80562",
         "42537",
         "Colon",
         "Stage IV-A ",
         "Visium HD, Sample P3 NAT",
         "v1_1_0"
        ],
        [
         "7",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX152",
         "TENX152.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Visium",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         "100.0",
         "55.0",
         "4269",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.2737239603302694",
         "40x",
         "76190",
         "49099",
         "Colon",
         null,
         "Visium CytAssist v2, Sample P2 CRC",
         "v1_1_0"
        ],
        [
         "8",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX149",
         "TENX149.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Xenium",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         null,
         null,
         "4009",
         "FFPE",
         "541",
         null,
         "0.2737768185611484",
         "0.2740308880582206",
         "40x",
         "70534",
         "60222",
         "Colon",
         "Stage II-A ",
         "Xenium In Situ, Sample P1 CRC",
         "v1_1_0"
        ],
        [
         "9",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX148",
         "TENX148.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Xenium",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         null,
         null,
         "4379",
         "FFPE",
         "541",
         null,
         "0.2737761889511267",
         "0.2737603571635976",
         "40x",
         "75184",
         "48582",
         "Colon",
         null,
         "Xenium In Situ, Sample P2 CRC",
         "v1_1_0"
        ],
        [
         "10",
         "Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high definition spatial profiling",
         "TENX147",
         "TENX147.tif",
         "Bowel",
         "Cancer",
         "COAD",
         "Homo sapiens",
         "Patient 1",
         "Xenium",
         "7/11/24",
         "Creative Commons Attribution",
         "https://www.biorxiv.org/content/10.1101/2024.06.04.597233v1.full",
         "https://www.10xgenomics.com/products/visium-hd-spatial-gene-expression/dataset-human-crc",
         null,
         null,
         "3997",
         "FFPE",
         "541",
         null,
         "0.2737753719614869",
         "0.2497022099084696",
         "40x",
         "71082",
         "66226",
         "Colon",
         "Stage IV-A ",
         "Xenium In Situ, Sample P5 CRC",
         "v1_1_0"
        ],
        [
         "13",
         "Visium HD Spatial Gene Expression Library, Human Pancreas (FFPE)",
         "TENX144",
         "TENX144.tif",
         "Pancreas",
         "Healthy",
         null,
         "Homo sapiens",
         null,
         "Visium HD",
         "6/11/24",
         "Creative Commons Attribution",
         null,
         "https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-human-pancreas",
         "2.0",
         "2.0",
         "1848",
         "FFPE",
         "18085",
         null,
         "0.2645833333333333",
         "0.2737638009601618",
         "40x",
         "51701",
         "51420",
         "Pancreas",
         null,
         null,
         "v1_1_0"
        ],
        [
         "14",
         "Preview Data: FFPE Human Lymph Node with 5K Pan Tissue and Pathways Panel",
         "TENX143",
         "TENX143.tif",
         "Lymphoid",
         "Diseased",
         null,
         "Homo sapiens",
         null,
         "Xenium",
         "5/28/24",
         "Creative Commons Attribution",
         null,
         "https://www.10xgenomics.com/datasets/preview-data-xenium-prime-gene-expression",
         null,
         null,
         "4889",
         "FFPE",
         "11094",
         null,
         "0.2737726437011389",
         "0.273951371",
         "40x",
         "28915",
         "30987",
         "Lymph node (Left Axilla)",
         "Reactive Lymph Node",
         null,
         "v1_1_0"
        ],
        [
         "19",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI884",
         "NCBI884.tif",
         "Lung",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 27",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "2504",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "25630",
         "28964",
         "Lung",
         "Control",
         "THD0008",
         "v1_1_0"
        ],
        [
         "20",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI883",
         "NCBI883.tif",
         "Lung",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 28",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1277",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17145",
         "16597",
         "Lung",
         "Control",
         "THD0011",
         "v1_1_0"
        ],
        [
         "21",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI882",
         "NCBI882.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 1",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1435",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17141",
         "20708",
         "Lung",
         "ILD More fibrotic",
         "TILD117MF",
         "v1_1_0"
        ],
        [
         "22",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI881",
         "NCBI881.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 2",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1392",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "19973",
         "16596",
         "Lung",
         "ILD Less fibrotic",
         "TILD117LF",
         "v1_1_0"
        ],
        [
         "23",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI880",
         "NCBI880.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 3",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1168",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17141",
         "16594",
         "Lung",
         "IPF Intermediate",
         "TILD175",
         "v1_1_0"
        ],
        [
         "24",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI879",
         "NCBI879.tif",
         "Lung",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 4",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1017",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "14312",
         "16598",
         "Lung",
         "Control",
         "VUHD069",
         "v1_1_0"
        ],
        [
         "25",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI876",
         "NCBI876.tif",
         "Lung",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 7",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1062",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "14301",
         "16598",
         "Lung",
         "Control",
         "VUHD116A",
         "v1_1_0"
        ],
        [
         "26",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI875",
         "NCBI875.tif",
         "Lung",
         "Healthy",
         null,
         "Homo sapiens",
         "Patient 7",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "999",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17132",
         "12491",
         "Lung",
         "Control",
         "VUHD116B",
         "v1_1_0"
        ],
        [
         "27",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI873",
         "NCBI873.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 9",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1484",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17134",
         "20714",
         "Lung",
         "IPF Less fibrotic",
         "VUILD102LF",
         "v1_1_0"
        ],
        [
         "28",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI870",
         "NCBI870.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 12",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1073",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17135",
         "16593",
         "Lung",
         "IPF More fibrotic",
         "VUILD105MF",
         "v1_1_0"
        ],
        [
         "29",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI867",
         "NCBI867.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 15",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "3499",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "28462",
         "33048",
         "Lung",
         "IPF Intermediate",
         "VUILD106",
         "v1_1_0"
        ],
        [
         "30",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI866",
         "NCBI866.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 16",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1604",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "19966",
         "20710",
         "Lung",
         "IPF More fibrotic",
         "VUILD107MF",
         "v1_1_0"
        ],
        [
         "31",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI865",
         "NCBI865.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 17",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "3062",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "28454",
         "28935",
         "Lung",
         "CTD-ILD Intermediate",
         "VUILD110",
         "v1_1_0"
        ],
        [
         "32",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI864",
         "NCBI864.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 18",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "2917",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "25624",
         "33047",
         "Lung",
         "cHP Intermediate",
         "VUILD115",
         "v1_1_0"
        ],
        [
         "33",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI861",
         "NCBI861.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 21",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1121",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17137",
         "16596",
         "Lung",
         "IPAF More fibrotic",
         "VUILD78MF",
         "v1_1_0"
        ],
        [
         "34",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI860",
         "NCBI860.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 22",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1282",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "19969",
         "16597",
         "Lung",
         "IPAF Less fibrotic",
         "VUILD78LF",
         "v1_1_0"
        ],
        [
         "35",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI859",
         "NCBI859.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 23",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1116",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17140",
         "16592",
         "Lung",
         "IPF More fibrotic",
         "VUILD91MF",
         "v1_1_0"
        ],
        [
         "36",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI858",
         "NCBI858.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 24",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "837",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "14306",
         "12484",
         "Lung",
         "IPF Less fibrotic",
         "VUILD91LF",
         "v1_1_0"
        ],
        [
         "37",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI857",
         "NCBI857.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 25",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1767",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "19964",
         "24815",
         "Lung",
         "Sarcoidosis More fibrotic",
         "VUILD96MF",
         "v1_1_0"
        ],
        [
         "38",
         "Image-based spatial transcriptomics identifies molecular niche dysregulation associated with distal lung remodeling in pulmonary fibrosis",
         "NCBI856",
         "NCBI856.tif",
         "Lung",
         "Diseased",
         null,
         "Homo sapiens",
         "Patient 26",
         "Xenium",
         "23-Dec-23",
         null,
         "https://www.biorxiv.org/content/10.1101/2023.12.15.571954v2",
         "https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE250346",
         null,
         null,
         "1607",
         "FFPE",
         "541",
         null,
         "0.2125000159375012",
         "0.2125",
         "40x",
         "17136",
         "24809",
         "Lung",
         "Sarcoidosis Less fibrotic",
         "VUILD96LF",
         "v1_1_0"
        ],
        [
         "39",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC142",
         "MISC142.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "1302",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4549588653433921",
         "40x",
         "18063",
         "18100",
         "right cardiac atrium",
         null,
         "HCAHeartST10238251",
         "v1_1_0"
        ],
        [
         "40",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC141",
         "MISC141.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "1702",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.455062497",
         "40x",
         "17890",
         "18098",
         "left cardiac atrium",
         null,
         "HCAHeartST10238252",
         "v1_1_0"
        ],
        [
         "41",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC140",
         "MISC140.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2168",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4552351838908369",
         "40x",
         "17802",
         "18044",
         "apical region of left ventricle",
         null,
         "HCAHeartST10238253",
         "v1_1_0"
        ],
        [
         "42",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC139",
         "MISC139.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2562",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4555808656036446",
         "40x",
         "17634",
         "17674",
         "heart right ventricle",
         null,
         "HCAHeartST10550730",
         "v1_1_0"
        ],
        [
         "43",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC138",
         "MISC138.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2199",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4554863862466527",
         "40x",
         "17532",
         "17473",
         "interventricular septum",
         null,
         "HCAHeartST10550732",
         "v1_1_0"
        ],
        [
         "44",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC136",
         "MISC136.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2090",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4418219026315251",
         "40x",
         "17955",
         "18186",
         "atrioventricular node",
         null,
         "HCAHeartST11290657",
         "v1_1_0"
        ],
        [
         "45",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC135",
         "MISC135.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2697",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4418630963595568",
         "40x",
         "17753",
         "18250",
         "atrioventricular node",
         null,
         "HCAHeartST11290658",
         "v1_1_0"
        ],
        [
         "46",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC134",
         "MISC134.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2338",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4417221454565067",
         "40x",
         "17772",
         "18231",
         "atrioventricular node",
         null,
         "HCAHeartST11290659",
         "v1_1_0"
        ],
        [
         "47",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC133",
         "MISC133.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3153",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4421340367360145",
         "40x",
         "17826",
         "18163",
         "atrioventricular node",
         null,
         "HCAHeartST11290662",
         "v1_1_0"
        ],
        [
         "48",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC132",
         "MISC132.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2843",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.442477394822145",
         "40x",
         "17861",
         "18159",
         "interventricular septum",
         null,
         "HCAHeartST11350375",
         "v1_1_0"
        ],
        [
         "49",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC131",
         "MISC131.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2807",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4424762584632593",
         "40x",
         "17721",
         "18204",
         "atrioventricular node",
         null,
         "HCAHeartST11350377",
         "v1_1_0"
        ],
        [
         "50",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC130",
         "MISC130.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3533",
         "FFPE",
         "17922",
         null,
         "31.7500635",
         "0.4553732173362386",
         "40x",
         "17222",
         "17629",
         "sinoatrial node",
         null,
         "HCAHeartST11702008",
         "v1_1_0"
        ],
        [
         "51",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC129",
         "MISC129.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3533",
         "FFPE",
         "17922",
         null,
         "31.7500635",
         "0.4553732173362386",
         "40x",
         "17243",
         "17604",
         "sinoatrial node",
         null,
         "HCAHeartST11702009",
         "v1_1_0"
        ],
        [
         "52",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC128",
         "MISC128.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3982",
         "FFPE",
         "17922",
         null,
         "31.7500635",
         "0.455062440221303",
         "40x",
         "17424",
         "17709",
         "sinoatrial node",
         null,
         "HCAHeartST11702010",
         "v1_1_0"
        ],
        [
         "53",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC127",
         "MISC127.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3802",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.503460189",
         "20x",
         "15610",
         "16028",
         "sinoatrial node",
         null,
         "HCAHeartST12992072",
         "v1_1_0"
        ],
        [
         "54",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC126",
         "MISC126.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "4521",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.5032456325843153",
         "20x",
         "15829",
         "16103",
         "atrioventricular node",
         null,
         "HCAHeartST12992074",
         "v1_1_0"
        ],
        [
         "55",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC125",
         "MISC125.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "3478",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.5033545976382061",
         "20x",
         "15987",
         "16236",
         "atrioventricular node",
         null,
         "HCAHeartST12992075",
         "v1_1_0"
        ],
        [
         "56",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC124",
         "MISC124.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2942",
         "Fresh Frozen",
         "33538",
         null,
         "63.500127",
         "0.5999988",
         "20x",
         "12997",
         "13330",
         "atrioventricular node",
         null,
         "HCAHeartST13228103",
         "v1_1_0"
        ]
       ],
       "shape": {
        "columns": 28,
        "rows": 649
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>id</th>\n",
       "      <th>image_filename</th>\n",
       "      <th>organ</th>\n",
       "      <th>disease_state</th>\n",
       "      <th>oncotree_code</th>\n",
       "      <th>species</th>\n",
       "      <th>patient</th>\n",
       "      <th>st_technology</th>\n",
       "      <th>data_publication_date</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_comment</th>\n",
       "      <th>pixel_size_um_embedded</th>\n",
       "      <th>pixel_size_um_estimated</th>\n",
       "      <th>magnification</th>\n",
       "      <th>fullres_px_width</th>\n",
       "      <th>fullres_px_height</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease_comment</th>\n",
       "      <th>subseries</th>\n",
       "      <th>hest_version_added</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FFPE Human Skin Primary Dermal Melanoma with 5...</td>\n",
       "      <td>TENX158</td>\n",
       "      <td>TENX158.tif</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>SKCM</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xenium</td>\n",
       "      <td>7/31/24</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.273777</td>\n",
       "      <td>0.273754</td>\n",
       "      <td>40x</td>\n",
       "      <td>18669</td>\n",
       "      <td>35787</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Primary Dermal Melanoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FFPE Human Prostate Adenocarcinoma with 5K Hum...</td>\n",
       "      <td>TENX157</td>\n",
       "      <td>TENX157.tif</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xenium</td>\n",
       "      <td>7/31/24</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.273772</td>\n",
       "      <td>0.273741</td>\n",
       "      <td>40x</td>\n",
       "      <td>25002</td>\n",
       "      <td>49976</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Characterization of immune cell populations in...</td>\n",
       "      <td>TENX156</td>\n",
       "      <td>TENX156.tif</td>\n",
       "      <td>Bowel</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>COAD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>Visium HD</td>\n",
       "      <td>7/11/24</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.264583</td>\n",
       "      <td>0.273802</td>\n",
       "      <td>40x</td>\n",
       "      <td>71106</td>\n",
       "      <td>58791</td>\n",
       "      <td>Colon</td>\n",
       "      <td>Stage II-A</td>\n",
       "      <td>Visium HD, Sample P1 CRC</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Characterization of immune cell populations in...</td>\n",
       "      <td>TENX155</td>\n",
       "      <td>TENX155.tif</td>\n",
       "      <td>Bowel</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>COAD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>Visium HD</td>\n",
       "      <td>7/11/24</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.264583</td>\n",
       "      <td>0.273874</td>\n",
       "      <td>40x</td>\n",
       "      <td>75250</td>\n",
       "      <td>48740</td>\n",
       "      <td>Colon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Visium HD, Sample P2 CRC</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Characterization of immune cell populations in...</td>\n",
       "      <td>TENX154</td>\n",
       "      <td>TENX154.tif</td>\n",
       "      <td>Bowel</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>COAD</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Patient 1</td>\n",
       "      <td>Visium HD</td>\n",
       "      <td>7/11/24</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.264583</td>\n",
       "      <td>0.273771</td>\n",
       "      <td>40x</td>\n",
       "      <td>72897</td>\n",
       "      <td>64370</td>\n",
       "      <td>Colon</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Visium HD, Sample P5 CRC</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1224</th>\n",
       "      <td>spatialLIBD</td>\n",
       "      <td>MISC5</td>\n",
       "      <td>MISC5.tif</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.727113</td>\n",
       "      <td>20x</td>\n",
       "      <td>13332</td>\n",
       "      <td>13332</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151672</td>\n",
       "      <td>v1_0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225</th>\n",
       "      <td>spatialLIBD</td>\n",
       "      <td>MISC4</td>\n",
       "      <td>MISC4.tif</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.726109</td>\n",
       "      <td>20x</td>\n",
       "      <td>13332</td>\n",
       "      <td>13332</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151673</td>\n",
       "      <td>v1_0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1226</th>\n",
       "      <td>spatialLIBD</td>\n",
       "      <td>MISC3</td>\n",
       "      <td>MISC3.tif</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.725124</td>\n",
       "      <td>20x</td>\n",
       "      <td>13332</td>\n",
       "      <td>13332</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151674</td>\n",
       "      <td>v1_0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1227</th>\n",
       "      <td>spatialLIBD</td>\n",
       "      <td>MISC2</td>\n",
       "      <td>MISC2.tif</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.726109</td>\n",
       "      <td>20x</td>\n",
       "      <td>13332</td>\n",
       "      <td>13332</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151675</td>\n",
       "      <td>v1_0_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1228</th>\n",
       "      <td>spatialLIBD</td>\n",
       "      <td>MISC1</td>\n",
       "      <td>MISC1.tif</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.726109</td>\n",
       "      <td>20x</td>\n",
       "      <td>13332</td>\n",
       "      <td>13332</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151676</td>\n",
       "      <td>v1_0_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>649 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          dataset_title       id  \\\n",
       "1     FFPE Human Skin Primary Dermal Melanoma with 5...  TENX158   \n",
       "2     FFPE Human Prostate Adenocarcinoma with 5K Hum...  TENX157   \n",
       "3     Characterization of immune cell populations in...  TENX156   \n",
       "4     Characterization of immune cell populations in...  TENX155   \n",
       "5     Characterization of immune cell populations in...  TENX154   \n",
       "...                                                 ...      ...   \n",
       "1224                                        spatialLIBD    MISC5   \n",
       "1225                                        spatialLIBD    MISC4   \n",
       "1226                                        spatialLIBD    MISC3   \n",
       "1227                                        spatialLIBD    MISC2   \n",
       "1228                                        spatialLIBD    MISC1   \n",
       "\n",
       "     image_filename     organ disease_state oncotree_code       species  \\\n",
       "1       TENX158.tif      Skin        Cancer          SKCM  Homo sapiens   \n",
       "2       TENX157.tif  Prostate        Cancer          PRAD  Homo sapiens   \n",
       "3       TENX156.tif     Bowel        Cancer          COAD  Homo sapiens   \n",
       "4       TENX155.tif     Bowel        Cancer          COAD  Homo sapiens   \n",
       "5       TENX154.tif     Bowel        Cancer          COAD  Homo sapiens   \n",
       "...             ...       ...           ...           ...           ...   \n",
       "1224      MISC5.tif     Brain       Healthy           NaN  Homo sapiens   \n",
       "1225      MISC4.tif     Brain       Healthy           NaN  Homo sapiens   \n",
       "1226      MISC3.tif     Brain       Healthy           NaN  Homo sapiens   \n",
       "1227      MISC2.tif     Brain       Healthy           NaN  Homo sapiens   \n",
       "1228      MISC1.tif     Brain       Healthy           NaN  Homo sapiens   \n",
       "\n",
       "        patient st_technology data_publication_date  ... treatment_comment  \\\n",
       "1           NaN        Xenium               7/31/24  ...               NaN   \n",
       "2           NaN        Xenium               7/31/24  ...               NaN   \n",
       "3     Patient 1     Visium HD               7/11/24  ...               NaN   \n",
       "4     Patient 1     Visium HD               7/11/24  ...               NaN   \n",
       "5     Patient 1     Visium HD               7/11/24  ...               NaN   \n",
       "...         ...           ...                   ...  ...               ...   \n",
       "1224                   Visium                   NaN  ...               NaN   \n",
       "1225                   Visium                   NaN  ...               NaN   \n",
       "1226                   Visium                   NaN  ...               NaN   \n",
       "1227                   Visium                   NaN  ...               NaN   \n",
       "1228                   Visium                   NaN  ...               NaN   \n",
       "\n",
       "     pixel_size_um_embedded pixel_size_um_estimated  magnification  \\\n",
       "1                  0.273777                0.273754            40x   \n",
       "2                  0.273772                0.273741            40x   \n",
       "3                  0.264583                0.273802            40x   \n",
       "4                  0.264583                0.273874            40x   \n",
       "5                  0.264583                0.273771            40x   \n",
       "...                     ...                     ...            ...   \n",
       "1224                    NaN                0.727113            20x   \n",
       "1225                    NaN                0.726109            20x   \n",
       "1226                    NaN                0.725124            20x   \n",
       "1227                    NaN                0.726109            20x   \n",
       "1228                    NaN                0.726109            20x   \n",
       "\n",
       "      fullres_px_width  fullres_px_height                          tissue  \\\n",
       "1                18669              35787                            Skin   \n",
       "2                25002              49976                        Prostate   \n",
       "3                71106              58791                           Colon   \n",
       "4                75250              48740                           Colon   \n",
       "5                72897              64370                           Colon   \n",
       "...                ...                ...                             ...   \n",
       "1224             13332              13332  dorsolateral prefrontal cortex   \n",
       "1225             13332              13332  dorsolateral prefrontal cortex   \n",
       "1226             13332              13332  dorsolateral prefrontal cortex   \n",
       "1227             13332              13332  dorsolateral prefrontal cortex   \n",
       "1228             13332              13332  dorsolateral prefrontal cortex   \n",
       "\n",
       "               disease_comment                 subseries  hest_version_added  \n",
       "1     Primary Dermal Melanoma                        NaN              v1_1_0  \n",
       "2                          NaN                       NaN              v1_1_0  \n",
       "3                  Stage II-A   Visium HD, Sample P1 CRC              v1_1_0  \n",
       "4                          NaN  Visium HD, Sample P2 CRC              v1_1_0  \n",
       "5                          NaN  Visium HD, Sample P5 CRC              v1_1_0  \n",
       "...                        ...                       ...                 ...  \n",
       "1224                       NaN                    151672              v1_0_0  \n",
       "1225                       NaN                    151673              v1_0_0  \n",
       "1226                       NaN                    151674              v1_0_0  \n",
       "1227                       NaN                    151675              v1_0_0  \n",
       "1228                       NaN                    151676              v1_0_0  \n",
       "\n",
       "[649 rows x 28 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bc9e7e6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_R_robust(data1, data2):\n",
    "    \"\"\"Robust correlation calculation handling edge cases\"\"\"\n",
    "    x1 = data1.X\n",
    "    x2 = data2.X\n",
    "    \n",
    "    print(f\"Data shapes: pred {x1.shape}, gt {x2.shape}\")\n",
    "    \n",
    "    if x1.shape != x2.shape:\n",
    "        print(\"ERROR: Shape mismatch between predicted and ground truth data\")\n",
    "        return np.array([]), np.array([])\n",
    "    \n",
    "    r_values = []\n",
    "    \n",
    "    for i in range(x1.shape[1]):  # For each gene\n",
    "        col1 = x1[:, i]\n",
    "        col2 = x2[:, i]\n",
    "        \n",
    "        # Check for constant columns (zero variance)\n",
    "        if np.var(col1) == 0 or np.var(col2) == 0:\n",
    "            r_values.append(0.0)\n",
    "            continue\n",
    "            \n",
    "        # Check for valid data points\n",
    "        mask = np.isfinite(col1) & np.isfinite(col2)\n",
    "        if mask.sum() < 2:\n",
    "            r_values.append(0.0)\n",
    "            continue\n",
    "            \n",
    "        try:\n",
    "            r, _ = pearsonr(col1[mask], col2[mask])\n",
    "            r_values.append(r if np.isfinite(r) else 0.0)\n",
    "        except:\n",
    "            r_values.append(0.0)\n",
    "    \n",
    "    return np.array(r_values), np.array([])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6ba0f189",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: MARS not in conversion dictionary.\n",
      "Error: CIR1 not in conversion dictionary.\n",
      "Error: PTRF not in conversion dictionary.\n",
      "Error: MLLT4 not in conversion dictionary.\n",
      "Error: C19orf24 not in conversion dictionary.\n",
      "Error: FAM213B not in conversion dictionary.\n",
      "Error: METTL12 not in conversion dictionary.\n",
      "Error: SLC9A3R2 not in conversion dictionary.\n",
      "Error: C4orf48 not in conversion dictionary.\n",
      "Error: SLC22A18 not in conversion dictionary.\n",
      "Error: C5orf38 not in conversion dictionary.\n",
      "Error: ATP5O not in conversion dictionary.\n",
      "Error: C19orf52 not in conversion dictionary.\n",
      "Error: KIAA1211L not in conversion dictionary.\n",
      "Error: C19orf60 not in conversion dictionary.\n",
      "Error: WDR34 not in conversion dictionary.\n",
      "Error: C12orf45 not in conversion dictionary.\n",
      "Error: KIAA1715 not in conversion dictionary.\n",
      "Error: SSSCA1 not in conversion dictionary.\n",
      "Error: MFSD7 not in conversion dictionary.\n",
      "Error: FAM173A not in conversion dictionary.\n",
      "Error: GLTSCR2 not in conversion dictionary.\n",
      "Error: C10orf54 not in conversion dictionary.\n",
      "Error: FAM102A not in conversion dictionary.\n",
      "Error: LINC00176 not in conversion dictionary.\n",
      "Error: VARS not in conversion dictionary.\n",
      "Error: FAM45A not in conversion dictionary.\n",
      "Error: ASNA1 not in conversion dictionary.\n",
      "Error: HSPB11 not in conversion dictionary.\n",
      "Error: ATP5H not in conversion dictionary.\n",
      "Error: KIAA0895L not in conversion dictionary.\n",
      "Error: KIAA1468 not in conversion dictionary.\n",
      "Error: C10orf10 not in conversion dictionary.\n",
      "Error: C6orf48 not in conversion dictionary.\n",
      "Error: FAM160B1 not in conversion dictionary.\n",
      "Error: UFD1L not in conversion dictionary.\n",
      "Error: AES not in conversion dictionary.\n",
      "HEST Dataset: 1 samples (test)\n",
      "['NCBI474']\n",
      "Using 785 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                           | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample NCBI474 has 758 common genes with the dataset\n",
      "Batch 0: patch shape = torch.Size([1, 598, 3, 112, 112])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 1/1 [00:02<00:00,  2.18s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final shapes - preds: (598, 785), ct: (598, 2), gt: (598, 785)\n",
      "Data shapes: pred (598, 785), gt (598, 785)\n",
      "Pearson Correlation: 0.0000\n",
      "Valid correlations: 785/785\n",
      "Error: MARS not in conversion dictionary.\n",
      "Error: CIR1 not in conversion dictionary.\n",
      "Error: PTRF not in conversion dictionary.\n",
      "Error: MLLT4 not in conversion dictionary.\n",
      "Error: C19orf24 not in conversion dictionary.\n",
      "Error: FAM213B not in conversion dictionary.\n",
      "Error: METTL12 not in conversion dictionary.\n",
      "Error: SLC9A3R2 not in conversion dictionary.\n",
      "Error: C4orf48 not in conversion dictionary.\n",
      "Error: SLC22A18 not in conversion dictionary.\n",
      "Error: C5orf38 not in conversion dictionary.\n",
      "Error: ATP5O not in conversion dictionary.\n",
      "Error: C19orf52 not in conversion dictionary.\n",
      "Error: KIAA1211L not in conversion dictionary.\n",
      "Error: C19orf60 not in conversion dictionary.\n",
      "Error: WDR34 not in conversion dictionary.\n",
      "Error: C12orf45 not in conversion dictionary.\n",
      "Error: KIAA1715 not in conversion dictionary.\n",
      "Error: SSSCA1 not in conversion dictionary.\n",
      "Error: MFSD7 not in conversion dictionary.\n",
      "Error: FAM173A not in conversion dictionary.\n",
      "Error: GLTSCR2 not in conversion dictionary.\n",
      "Error: C10orf54 not in conversion dictionary.\n",
      "Error: FAM102A not in conversion dictionary.\n",
      "Error: LINC00176 not in conversion dictionary.\n",
      "Error: VARS not in conversion dictionary.\n",
      "Error: FAM45A not in conversion dictionary.\n",
      "Error: ASNA1 not in conversion dictionary.\n",
      "Error: HSPB11 not in conversion dictionary.\n",
      "Error: ATP5H not in conversion dictionary.\n",
      "Error: KIAA0895L not in conversion dictionary.\n",
      "Error: KIAA1468 not in conversion dictionary.\n",
      "Error: C10orf10 not in conversion dictionary.\n",
      "Error: C6orf48 not in conversion dictionary.\n",
      "Error: FAM160B1 not in conversion dictionary.\n",
      "Error: UFD1L not in conversion dictionary.\n",
      "Error: AES not in conversion dictionary.\n",
      "HEST Dataset: 1 samples (test)\n",
      "['NCBI539']\n",
      "Using 785 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                           | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample NCBI539 has 771 common genes with the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 1/1 [00:02<00:00,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch 0: patch shape = torch.Size([1, 475, 3, 112, 112])\n",
      "Final shapes - preds: (475, 785), ct: (475, 2), gt: (475, 785)\n",
      "Data shapes: pred (475, 785), gt (475, 785)\n",
      "Pearson Correlation: 0.0000\n",
      "Valid correlations: 785/785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: MARS not in conversion dictionary.\n",
      "Error: CIR1 not in conversion dictionary.\n",
      "Error: PTRF not in conversion dictionary.\n",
      "Error: MLLT4 not in conversion dictionary.\n",
      "Error: C19orf24 not in conversion dictionary.\n",
      "Error: FAM213B not in conversion dictionary.\n",
      "Error: METTL12 not in conversion dictionary.\n",
      "Error: SLC9A3R2 not in conversion dictionary.\n",
      "Error: C4orf48 not in conversion dictionary.\n",
      "Error: SLC22A18 not in conversion dictionary.\n",
      "Error: C5orf38 not in conversion dictionary.\n",
      "Error: ATP5O not in conversion dictionary.\n",
      "Error: C19orf52 not in conversion dictionary.\n",
      "Error: KIAA1211L not in conversion dictionary.\n",
      "Error: C19orf60 not in conversion dictionary.\n",
      "Error: WDR34 not in conversion dictionary.\n",
      "Error: C12orf45 not in conversion dictionary.\n",
      "Error: KIAA1715 not in conversion dictionary.\n",
      "Error: SSSCA1 not in conversion dictionary.\n",
      "Error: MFSD7 not in conversion dictionary.\n",
      "Error: FAM173A not in conversion dictionary.\n",
      "Error: GLTSCR2 not in conversion dictionary.\n",
      "Error: C10orf54 not in conversion dictionary.\n",
      "Error: FAM102A not in conversion dictionary.\n",
      "Error: LINC00176 not in conversion dictionary.\n",
      "Error: VARS not in conversion dictionary.\n",
      "Error: FAM45A not in conversion dictionary.\n",
      "Error: ASNA1 not in conversion dictionary.\n",
      "Error: HSPB11 not in conversion dictionary.\n",
      "Error: ATP5H not in conversion dictionary.\n",
      "Error: KIAA0895L not in conversion dictionary.\n",
      "Error: KIAA1468 not in conversion dictionary.\n",
      "Error: C10orf10 not in conversion dictionary.\n",
      "Error: C6orf48 not in conversion dictionary.\n",
      "Error: FAM160B1 not in conversion dictionary.\n",
      "Error: UFD1L not in conversion dictionary.\n",
      "Error: AES not in conversion dictionary.\n",
      "HEST Dataset: 1 samples (test)\n",
      "['MISC61']\n",
      "Using 785 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                           | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample MISC61 has 755 common genes with the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 1/1 [00:04<00:00,  4.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch 0: patch shape = torch.Size([1, 2047, 3, 112, 112])\n",
      "Final shapes - preds: (2047, 785), ct: (2047, 2), gt: (2047, 785)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data shapes: pred (2047, 785), gt (2047, 785)\n",
      "Pearson Correlation: 0.0405\n",
      "Valid correlations: 785/785\n",
      "Error: MARS not in conversion dictionary.\n",
      "Error: CIR1 not in conversion dictionary.\n",
      "Error: PTRF not in conversion dictionary.\n",
      "Error: MLLT4 not in conversion dictionary.\n",
      "Error: C19orf24 not in conversion dictionary.\n",
      "Error: FAM213B not in conversion dictionary.\n",
      "Error: METTL12 not in conversion dictionary.\n",
      "Error: SLC9A3R2 not in conversion dictionary.\n",
      "Error: C4orf48 not in conversion dictionary.\n",
      "Error: SLC22A18 not in conversion dictionary.\n",
      "Error: C5orf38 not in conversion dictionary.\n",
      "Error: ATP5O not in conversion dictionary.\n",
      "Error: C19orf52 not in conversion dictionary.\n",
      "Error: KIAA1211L not in conversion dictionary.\n",
      "Error: C19orf60 not in conversion dictionary.\n",
      "Error: WDR34 not in conversion dictionary.\n",
      "Error: C12orf45 not in conversion dictionary.\n",
      "Error: KIAA1715 not in conversion dictionary.\n",
      "Error: SSSCA1 not in conversion dictionary.\n",
      "Error: MFSD7 not in conversion dictionary.\n",
      "Error: FAM173A not in conversion dictionary.\n",
      "Error: GLTSCR2 not in conversion dictionary.\n",
      "Error: C10orf54 not in conversion dictionary.\n",
      "Error: FAM102A not in conversion dictionary.\n",
      "Error: LINC00176 not in conversion dictionary.\n",
      "Error: VARS not in conversion dictionary.\n",
      "Error: FAM45A not in conversion dictionary.\n",
      "Error: ASNA1 not in conversion dictionary.\n",
      "Error: HSPB11 not in conversion dictionary.\n",
      "Error: ATP5H not in conversion dictionary.\n",
      "Error: KIAA0895L not in conversion dictionary.\n",
      "Error: KIAA1468 not in conversion dictionary.\n",
      "Error: C10orf10 not in conversion dictionary.\n",
      "Error: C6orf48 not in conversion dictionary.\n",
      "Error: FAM160B1 not in conversion dictionary.\n",
      "Error: UFD1L not in conversion dictionary.\n",
      "Error: AES not in conversion dictionary.\n",
      "HEST Dataset: 1 samples (test)\n",
      "['SPA102']\n",
      "Using 785 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                           | 0/1 [00:02<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample SPA102 has 739 common genes with the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "np.str_('HPS6')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[21], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m hestset \u001b[38;5;241m=\u001b[39m pk_load(fold, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtest\u001b[39m\u001b[38;5;124m'\u001b[39m, sample_ids\u001b[38;5;241m=\u001b[39m[sample_id], dataset\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhest1k\u001b[39m\u001b[38;5;124m'\u001b[39m, flatten\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, adj\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, ori\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, prune\u001b[38;5;241m=\u001b[39mprune)\n\u001b[1;32m      4\u001b[0m hest_loader \u001b[38;5;241m=\u001b[39m DataLoader(hestset, batch_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, num_workers\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, shuffle\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m----> 5\u001b[0m pred, gt \u001b[38;5;241m=\u001b[39m \u001b[43mtest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhest_loader\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcuda\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# Use robust correlation\u001b[39;00m\n\u001b[1;32m      7\u001b[0m R, _ \u001b[38;5;241m=\u001b[39m get_R_robust(pred, gt)\n",
      "File \u001b[0;32m/work/bose_lab/Jamie/Summer/models/Hist2ST-Fork/predict.py:58\u001b[0m, in \u001b[0;36mtest\u001b[0;34m(model, test, device)\u001b[0m\n\u001b[1;32m     56\u001b[0m loss\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[0;32m---> 58\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, (patch, position, exp, adj, \u001b[38;5;241m*\u001b[39m_, center) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(tqdm(test)):\n\u001b[1;32m     59\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBatch \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: patch shape = \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpatch\u001b[38;5;241m.\u001b[39mshape\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     60\u001b[0m         patch, position, adj \u001b[38;5;241m=\u001b[39m patch\u001b[38;5;241m.\u001b[39mto(device), position\u001b[38;5;241m.\u001b[39mto(device), adj\u001b[38;5;241m.\u001b[39mto(device)\u001b[38;5;241m.\u001b[39msqueeze(\u001b[38;5;241m0\u001b[39m)\n",
      "File \u001b[0;32m~/software/miniforge3/envs/hist2st-env/lib/python3.9/site-packages/tqdm/std.py:1181\u001b[0m, in \u001b[0;36mtqdm.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1178\u001b[0m time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_time\n\u001b[1;32m   1180\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1181\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m obj \u001b[38;5;129;01min\u001b[39;00m iterable:\n\u001b[1;32m   1182\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m obj\n\u001b[1;32m   1183\u001b[0m         \u001b[38;5;66;03m# Update and possibly print the progressbar.\u001b[39;00m\n\u001b[1;32m   1184\u001b[0m         \u001b[38;5;66;03m# Note: does not call self.update(1) for speed optimisation.\u001b[39;00m\n",
      "File \u001b[0;32m~/software/miniforge3/envs/hist2st-env/lib/python3.9/site-packages/torch/utils/data/dataloader.py:701\u001b[0m, in \u001b[0;36m_BaseDataLoaderIter.__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    698\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sampler_iter \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    699\u001b[0m     \u001b[38;5;66;03m# TODO(https://github.com/pytorch/pytorch/issues/76750)\u001b[39;00m\n\u001b[1;32m    700\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reset()  \u001b[38;5;66;03m# type: ignore[call-arg]\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_next_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_num_yielded \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_dataset_kind \u001b[38;5;241m==\u001b[39m _DatasetKind\u001b[38;5;241m.\u001b[39mIterable\n\u001b[1;32m    705\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_IterableDataset_len_called \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_num_yielded \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_IterableDataset_len_called\n\u001b[1;32m    707\u001b[0m ):\n",
      "File \u001b[0;32m~/software/miniforge3/envs/hist2st-env/lib/python3.9/site-packages/torch/utils/data/dataloader.py:757\u001b[0m, in \u001b[0;36m_SingleProcessDataLoaderIter._next_data\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    755\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m_next_data\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    756\u001b[0m     index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_next_index()  \u001b[38;5;66;03m# may raise StopIteration\u001b[39;00m\n\u001b[0;32m--> 757\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dataset_fetcher\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfetch\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# may raise StopIteration\u001b[39;00m\n\u001b[1;32m    758\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pin_memory:\n\u001b[1;32m    759\u001b[0m         data \u001b[38;5;241m=\u001b[39m _utils\u001b[38;5;241m.\u001b[39mpin_memory\u001b[38;5;241m.\u001b[39mpin_memory(data, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_pin_memory_device)\n",
      "File \u001b[0;32m~/software/miniforge3/envs/hist2st-env/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py:52\u001b[0m, in \u001b[0;36m_MapDatasetFetcher.fetch\u001b[0;34m(self, possibly_batched_index)\u001b[0m\n\u001b[1;32m     50\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdataset\u001b[38;5;241m.\u001b[39m__getitems__(possibly_batched_index)\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 52\u001b[0m         data \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdataset[idx] \u001b[38;5;28;01mfor\u001b[39;00m idx \u001b[38;5;129;01min\u001b[39;00m possibly_batched_index]\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     54\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdataset[possibly_batched_index]\n",
      "File \u001b[0;32m~/software/miniforge3/envs/hist2st-env/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py:52\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     50\u001b[0m         data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdataset\u001b[38;5;241m.\u001b[39m__getitems__(possibly_batched_index)\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 52\u001b[0m         data \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdataset\u001b[49m\u001b[43m[\u001b[49m\u001b[43midx\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m idx \u001b[38;5;129;01min\u001b[39;00m possibly_batched_index]\n\u001b[1;32m     53\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     54\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdataset[possibly_batched_index]\n",
      "File \u001b[0;32m/work/bose_lab/Jamie/Summer/models/Hist2ST-Fork/dataset.py:496\u001b[0m, in \u001b[0;36mViT_HEST1K.__getitem__\u001b[0;34m(self, idx)\u001b[0m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;66;03m# Reorder columns to match gene_list order\u001b[39;00m\n\u001b[1;32m    495\u001b[0m gene_order \u001b[38;5;241m=\u001b[39m {gene: idx \u001b[38;5;28;01mfor\u001b[39;00m idx, gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(common_genes \u001b[38;5;241m+\u001b[39m missing_genes)}\n\u001b[0;32m--> 496\u001b[0m col_order \u001b[38;5;241m=\u001b[39m [gene_order[gene] \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgene_list]\n\u001b[1;32m    497\u001b[0m exps \u001b[38;5;241m=\u001b[39m exps[:, col_order]\n\u001b[1;32m    499\u001b[0m \u001b[38;5;66;03m#TODO: they normalized and log-transformed the data in the HER2ST dataset, should we do that here?\u001b[39;00m\n",
      "File \u001b[0;32m/work/bose_lab/Jamie/Summer/models/Hist2ST-Fork/dataset.py:496\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;66;03m# Reorder columns to match gene_list order\u001b[39;00m\n\u001b[1;32m    495\u001b[0m gene_order \u001b[38;5;241m=\u001b[39m {gene: idx \u001b[38;5;28;01mfor\u001b[39;00m idx, gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(common_genes \u001b[38;5;241m+\u001b[39m missing_genes)}\n\u001b[0;32m--> 496\u001b[0m col_order \u001b[38;5;241m=\u001b[39m [\u001b[43mgene_order\u001b[49m\u001b[43m[\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgene_list]\n\u001b[1;32m    497\u001b[0m exps \u001b[38;5;241m=\u001b[39m exps[:, col_order]\n\u001b[1;32m    499\u001b[0m \u001b[38;5;66;03m#TODO: they normalized and log-transformed the data in the HER2ST dataset, should we do that here?\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: np.str_('HPS6')"
     ]
    }
   ],
   "source": [
    "correlations = []\n",
    "for sample_id in sample_ids:\n",
    "    hestset = pk_load(fold, 'test', sample_ids=[sample_id], dataset='hest1k', flatten=False, adj=True, ori=True, prune=prune)\n",
    "    hest_loader = DataLoader(hestset, batch_size=1, num_workers=0, shuffle=False)\n",
    "    pred, gt = test(model, hest_loader,'cuda')\n",
    "    # Use robust correlation\n",
    "    R, _ = get_R_robust(pred, gt)\n",
    "    print(f'Pearson Correlation: {np.nanmean(R):.4f}')\n",
    "    print(f'Valid correlations: {np.sum(~np.isnan(R))}/{len(R)}')\n",
    "    correlations.append([sample_id, np.nanmean(R)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3131911",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['dataset_title', 'id', 'image_filename', 'organ', 'disease_state',\n",
       "       'oncotree_code', 'species', 'patient', 'st_technology',\n",
       "       'data_publication_date', 'license', 'study_link', 'download_page_link1',\n",
       "       'inter_spot_dist', 'spot_diameter', 'spots_under_tissue',\n",
       "       'preservation_method', 'nb_genes', 'treatment_comment',\n",
       "       'pixel_size_um_embedded', 'pixel_size_um_estimated', 'magnification',\n",
       "       'fullres_px_width', 'fullres_px_height', 'tissue', 'disease_comment',\n",
       "       'subseries', 'hest_version_added'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82c4f1f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "dataset_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "image_filename",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "organ",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "disease_state",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "oncotree_code",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "species",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "patient",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "st_technology",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "data_publication_date",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "license",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_link",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "download_page_link1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "inter_spot_dist",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "spot_diameter",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "spots_under_tissue",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "preservation_method",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "nb_genes",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "treatment_comment",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "pixel_size_um_embedded",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "pixel_size_um_estimated",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "magnification",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fullres_px_width",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "fullres_px_height",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "tissue",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "disease_comment",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "subseries",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "hest_version_added",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "3b1c2810-0222-4f22-8a3b-5c2784a62948",
       "rows": [
        [
         "42",
         "Spatially resolved multiomics of human cardiac niches",
         "MISC139",
         "MISC139.tif",
         "Heart",
         "Healthy",
         null,
         "Homo sapiens",
         "   ",
         "Visium",
         null,
         "Creative Commons Attribution 4.0 International",
         "https://www.nature.com/articles/s41586-023-06311-1#data-availability",
         "https://www.heartcellatlas.org/",
         "100.0",
         "55.0",
         "2562",
         "Fresh Frozen",
         "33538",
         null,
         "31.7500635",
         "0.4555808656036446",
         "40x",
         "17634",
         "17674",
         "heart right ventricle",
         null,
         "HCAHeartST10550730",
         "v1_1_0"
        ]
       ],
       "shape": {
        "columns": 28,
        "rows": 1
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>id</th>\n",
       "      <th>image_filename</th>\n",
       "      <th>organ</th>\n",
       "      <th>disease_state</th>\n",
       "      <th>oncotree_code</th>\n",
       "      <th>species</th>\n",
       "      <th>patient</th>\n",
       "      <th>st_technology</th>\n",
       "      <th>data_publication_date</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_comment</th>\n",
       "      <th>pixel_size_um_embedded</th>\n",
       "      <th>pixel_size_um_estimated</th>\n",
       "      <th>magnification</th>\n",
       "      <th>fullres_px_width</th>\n",
       "      <th>fullres_px_height</th>\n",
       "      <th>tissue</th>\n",
       "      <th>disease_comment</th>\n",
       "      <th>subseries</th>\n",
       "      <th>hest_version_added</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Spatially resolved multiomics of human cardiac...</td>\n",
       "      <td>MISC139</td>\n",
       "      <td>MISC139.tif</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td></td>\n",
       "      <td>Visium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.750063</td>\n",
       "      <td>0.455581</td>\n",
       "      <td>40x</td>\n",
       "      <td>17634</td>\n",
       "      <td>17674</td>\n",
       "      <td>heart right ventricle</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HCAHeartST10550730</td>\n",
       "      <td>v1_1_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        dataset_title       id image_filename  \\\n",
       "42  Spatially resolved multiomics of human cardiac...  MISC139    MISC139.tif   \n",
       "\n",
       "    organ disease_state oncotree_code       species patient st_technology  \\\n",
       "42  Heart       Healthy           NaN  Homo sapiens                Visium   \n",
       "\n",
       "   data_publication_date  ... treatment_comment pixel_size_um_embedded  \\\n",
       "42                   NaN  ...               NaN              31.750063   \n",
       "\n",
       "   pixel_size_um_estimated  magnification  fullres_px_width  \\\n",
       "42                0.455581            40x             17634   \n",
       "\n",
       "    fullres_px_height                 tissue  disease_comment  \\\n",
       "42              17674  heart right ventricle              NaN   \n",
       "\n",
       "             subseries  hest_version_added  \n",
       "42  HCAHeartST10550730              v1_1_0  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df[meta_df['id']=='MISC139']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d275198",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n"
     ]
    }
   ],
   "source": [
    "import anndata\n",
    "ad = anndata.read_h5ad(os.path.join(hest_path, \"st\", \"MISC139.h5ad\"))\n",
    "print(type(ad.var_names[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f270b151",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson Correlation: nan\n"
     ]
    }
   ],
   "source": [
    "# Clean the data by replacing NaN/Inf values\n",
    "def clean_data(adata):\n",
    "    \"\"\"Remove NaN and Inf values from AnnData object\"\"\"\n",
    "    X = adata.X.copy()\n",
    "    # Replace NaN with 0 or mean\n",
    "    X = np.nan_to_num(X, nan=0.0, posinf=0.0, neginf=0.0)\n",
    "    # Or use median imputation\n",
    "    # mask = np.isnan(X) | np.isinf(X)\n",
    "    # X[mask] = np.nanmedian(X)\n",
    "    adata_clean = ad.AnnData(X)\n",
    "    adata_clean.obsm = adata.obsm.copy()\n",
    "    return adata_clean\n",
    "\n",
    "# Clean both datasets before correlation\n",
    "pred_clean = clean_data(pred)\n",
    "\n",
    "R = get_R(pred_clean, gt)[0]\n",
    "print('Pearson Correlation:', np.nanmean(R))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f27dd627",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking pred data:\n",
      "  Shape: (598, 785)\n",
      "  Has NaN: True\n",
      "  Has Inf: False\n",
      "  Min value: nan\n",
      "  Max value: nan\n",
      "Checking gt data:\n",
      "  Shape: (598, 785)\n",
      "  Has NaN: False\n",
      "  Has Inf: False\n",
      "  Min value: 0.0\n",
      "  Max value: 4.0000433921813965\n"
     ]
    }
   ],
   "source": [
    "# Before calling get_R, inspect your data\n",
    "print(\"Checking pred data:\")\n",
    "print(f\"  Shape: {pred.X.shape}\")\n",
    "print(f\"  Has NaN: {np.isnan(pred.X).any()}\")\n",
    "print(f\"  Has Inf: {np.isinf(pred.X).any()}\")\n",
    "print(f\"  Min value: {np.nanmin(pred.X)}\")\n",
    "print(f\"  Max value: {np.nanmax(pred.X)}\")\n",
    "\n",
    "print(\"Checking gt data:\")\n",
    "print(f\"  Shape: {gt.X.shape}\")\n",
    "print(f\"  Has NaN: {np.isnan(gt.X).any()}\")\n",
    "print(f\"  Has Inf: {np.isinf(gt.X).any()}\")\n",
    "print(f\"  Min value: {np.nanmin(gt.X)}\")\n",
    "print(f\"  Max value: {np.nanmax(gt.X)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4266cce3-6bc7-4f5b-81d5-d03d3ae089c4",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson Correlation: 0.2887870599966082\n",
      "ARI: 0.431\n"
     ]
    }
   ],
   "source": [
    "R=get_R(pred,gt)[0]\n",
    "print('Pearson Correlation:',np.nanmean(R))\n",
    "\n",
    "\n",
    "# clus,ARI=cluster(pred,label)\n",
    "# print('ARI:',ARI)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
